packages/workspace/src/browser/workspace-service.ts

```diff
@@ -93,31 +157,154 @@ export class WorkspaceService implements FrontendApplicationContribution {
+                    this._workspaceConfigFile = await this.fileSystem.getFileStat(configPath);
+                } else {
+                    this._workspaceConfigFile = undefined;
+                }
+                this._workspaceRootChange.workspaceConfigFile = this._workspaceConfigFile;
+                this.startWatch(valid);
+                this.onWorkspaceRootChangeEmitter.fire(this._workspaceRootChange);
+            }
+            return Promise.resolve();
+        }
+        throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootToAdd}.`);
+    }
+
+    /**
+     * Removes folder(s) from workspace.
+     */
+    async removeFolders(uris: URI[]): Promise<void> {
+        if (!this.opened) {
+            throw new Error('Folder cannot be removed as there is no active folder in the current workspace.');
+        }
+        const validUris: string[] = [];
+        for (const uri of uris.map(u => u.toString())) {
+            const valid = await this.toValidRoot(uri);
+            if (valid) {
+                validUris.push(uri);
+            }
+        }
+        if (validUris.length > 0) {
+            if (this._root) {
+                const rootIndex = validUris.findIndex(uri => uri === this._root!.uri);
+                let rootName = '';
+                if (rootIndex >= 0) { // first root cannot be removed from the workspace
+                    validUris.splice(rootIndex, 1);
+                    rootName = uris[rootIndex].displayName;
+                    uris.splice(rootIndex, 1);
+                }
+
+                const messageContainer = document.createElement('div');
+                const hasRootFoldersToRemove = validUris.length > 0;
+                const isFirstRootFolderSelected = rootIndex >= 0;
+                if (hasRootFoldersToRemove) {
+                    messageContainer.textContent = 'Do you really want to remove the following roots and their subfolders from the workspace?';
+                    const list = document.createElement('ul');
+                    list.style.listStyleType = 'none';
+                    uris.forEach(u => {
+                        const listItem = document.createElement('li');
+                        listItem.textContent = u.displayName;
+                        list.appendChild(listItem);
+                    });
+                    messageContainer.appendChild(list);
+                }
+                if (isFirstRootFolderSelected) {
+                    const firstRootMsg = document.createElement('div');
+                    const firstRootNameDiv = document.createElement('strong');
+                    firstRootNameDiv.textContent = rootName;
+                    firstRootMsg.appendChild(firstRootNameDiv);
+                    const remainder = document.createElement('span');
+                    remainder.textContent = ' and its subfolders cannot be removed as it is the first root folder of the workspace.';
```

Comment:
As discussed, it may no longer be necessary to treat the first root folder in a special way, at least as far as removal goes.